### PR TITLE
[JENKINS-45864] Add SCM instance attribute to ChangeLogSet

### PR DESCRIPTION
**Pull request by thatrevguy**
See [JENKINS-45864](https://issues.jenkins-ci.org/browse/JENKINS-45864).

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by daniel-beck** [comment]
@abayer Didn't you recently build something related?

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: `@since TODO` is better

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: "... to support non-Pipeline jobs".

Ideally we need a `@RequireOverride` annotation to make these cases explicit during the compilation/static analysis

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: Makes sense to add Javadoc and to document why

File: core/src/main/java/hudson/scm/ChangeLogSet.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogSet.java]: Add `@CheckForNull` annotation

File: core/src/main/java/hudson/scm/ChangeLogSet.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogSet.java]: Use `this(run, browser, null)` to avoid code duplication

File: core/src/main/java/hudson/scm/ChangeLogSet.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogSet.java]: Add `@CheckForNull`

File: core/src/main/java/hudson/scm/ChangeLogSet.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogSet.java]: Nice2have, document parameters, add argument annotations

**Review by oleg-nenashev — COMMENTED** [review]
The change looks good to me, my comments are mostly about better diagnosability and static analysis

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: I would just say, `@deprecated this method should not be used directly anymore, use the {@link ...} method instead`. Should not override annotation may be helpful as well (`@Restricted(ProtectedExternally.class) `)

Comment (hathoward) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: A little confused on how AbstractBuild will behave when it runs [this](https://github.com/hathoward/jenkins/blob/13f25160d4f662641a2b449315a890e0207d376d/core/src/main/java/hudson/model/AbstractBuild.java#L861).

If a SCM plugin only overrides the [Run based parse function](https://github.com/hathoward/jenkins/blob/13f25160d4f662641a2b449315a890e0207d376d/core/src/main/java/hudson/scm/ChangeLogParser.java#L47) will AbstractBuild run it or still revert to [AbstractBuild based parse function](https://github.com/hathoward/jenkins/blob/13f25160d4f662641a2b449315a890e0207d376d/core/src/main/java/hudson/scm/ChangeLogParser.java#L72)?

Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: This is a good point. Maybe it makes sense to name the new method as `parseRun()` to prevent ambiguity. Then all internal calls may be safely moved to the new method

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (daniel-beck) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: Method renamed, also AME below needs adapting.

File: core/src/main/java/hudson/scm/ChangeLogParser.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/ChangeLogParser.java]: I would rather keep it as just "parse()" since it is not `Run`. Renaming both methods didn't help to avoid ambiguity I was referencing before

**Comment by oleg-nenashev** [comment]
Would be nice to get it in the weekly so it would have a chance to get into the next LTS baseline

**Review by oleg-nenashev — APPROVED** [review]

**Comment by daniel-beck** [comment]
Would prefer this gets a proper review to ensure the additions are sane.

**Comment by oleg-nenashev** [comment]
@daniel-beck put it on hold

**Comment by oleg-nenashev** [comment]
@daniel-beck seems it's going to miss the weekly. Would be great if you find some time to review it before the next one

**Comment by hathoward** [comment]
Hello, anything else needed from me at this point?

**Comment by daniel-beck** [comment]
@hathoward No, thanks for checking in. Jenkins World has been sucking up our time the last few weeks, so this has been delayed.

**Review by oleg-nenashev — APPROVED** [review]
The updated version LGTM, needs re-review by @daniel-beck

**Comment by daniel-beck** [comment]
Would be more comfortable if @jglick could take a look at this.

**Review by jglick — CHANGES_REQUESTED** [review]
Does not seem like a good idea. Adding a new persistent field to `ChangeLogSet` may or may not be desirable (would need to consider the XStream-serialized form). At least, on its own it does not address the stated goal

> a reliable way to identify correct remote URL for a ChangeLogSet instance

which based on the references to the `git` plugin I presume refers to Git changelogs. It is also unclear from JIRA what kind of report is needed—in the UI? Via Stapler exported `…/api/json`? Solely via Java API?

In any event there is no apparent downstream PR showing how this would be used, by `git-plugin`, or the reporter’s

> plugin for handling deployment COs

Certainly it would need at a minimum some kind of new call for Pipeline, thus a PR using timestamped snapshots in `workflow-job-plugin`, and some sort of functional test demonstrating the newly available information being used successfully.

It is not even obvious to me that any core API change is needed here at all. There is nothing preventing an implementation of `SCM.createChangeLogParser` from retaining some configuration fields from the `SCM`, such as a “remote URL”. The fact that `GitSCM` does not do so currently is presumably because no one cared before.

Whether any newly added getter in `GitChangeSetList` or `GitChangeSet` (unclear which) would need to `@Override` some API method in core would depend on whether there is a meaningful common definition of “remote URL”. Subversion, for example, uses URLs for everything, but the “repository” is just a server base URL, whereas projects typically operate on some subdirectory; and its changelog entries are able to specify both relative and absolute paths.

**Comment by jglick** [comment]
> Adding a new persistent field to `ChangeLogSet` may or may not be desirable

Ignore, it is not serializable.

**Comment by hathoward** [comment]
Thank you for taking a look at this.
> It is also unclear from JIRA what kind of report is needed—in the UI? Via Stapler exported …/api/json? Solely via Java API?

We have a plugin that generates and manages change orders in Service Now. Report is change order itself with contents including information about the change sets that triggered CICD pipeline. Reliably generating working URLs from change sets has been difficult when a pipeline performs multiple checkouts.
> In any event there is no apparent downstream PR showing how this would be used, by git-plugin

I have not figured out how to build a plugin against a non upstream Jenkins core build. If there is a trivial way of doing that it'd be much appreciated to learn.

>It is not even obvious to me that any core API change is needed here at all. There is nothing preventing an implementation of SCM.createChangeLogParser from retaining some configuration fields from the SCM, such as a “remote URL”. The fact that GitSCM does not do so currently is presumably because no one cared before.

This looks very much doable. I'll investigate and close this PR if I get a working PR for `git-plugin`.

**Comment by daniel-beck** [comment]
@hathoward 

> I have not figured out how to build a plugin against a non upstream Jenkins core build. If there is a trivial way of doing that it'd be much appreciated to learn.

Any registered user (accounts.jenkins.io) can upload snapshot builds of anything to repo.jenkins-ci.org (`mvn deploy`). These will get a unique timestamp on upload for unambiguity. You can specify that version as `jenkins.version` property in plugin POMs.

**Comment by batmat** [comment]
Hello @thatrevguy! do you plan to get back address comments here?

I am going to propose this PR for closing, and will close it if no action is taken in the next days or so. 
Please feel free to comment anytime, even after it would be closed, so we reopen if you'd like to work on finishing it.

Thanks!

**Pull request closed without merging**
